subsys/net/lib/http/http_server.c

```diff
@@ -0,0 +1,1205 @@
+static char http_response[CONFIG_NET_HTTP_SERVER_MAX_RESPONSE_SIZE];
+static const char *preface = "PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n";
+
+static struct http_parser_settings parserSettings;
+static struct http_parser parser;
+
+static struct arithmetic_result results[POST_REQUEST_STORAGE_LIMIT];
+static int results_count = 1;
+
+static unsigned char settings_frame[9] = {
+	0x00, 0x00, 0x00,        /* Length */
+	0x04,                    /* Type: 0x04 - setting frames for config or acknowledgment */
+	0x00,                    /* Flags: 0x00 - unused flags */
+	0x00, 0x00, 0x00, 0x00}; /* Reserved, Stream Identifier: 0x00 - overall connection */
+
+static unsigned char settings_ack[9] = {
+	0x00, 0x00, 0x00,        /* Length */
+	0x04,                    /* Type: 0x04 - setting frames for config or acknowledgment */
+	0x01,                    /* Flags: 0x01 - ACK */
+	0x00, 0x00, 0x00, 0x00}; /* Reserved, Stream Identifier */
+
+static const char content_404[] = {
+#ifdef INCLUDE_HTML_CONTENT
+#include "not_found_page.html.gz.inc"
+#endif
+};
+
+#define INVALID_SOCK -1
+
+bool has_upgrade_header = true;
+
+int http_server_init(struct http_server_ctx *ctx)
+{
+	int i = 1;
+	uint8_t af;
+	socklen_t len;
+	struct sockaddr_in6 addr_storage;
+	const union {
+		struct sockaddr *addr;
+		struct sockaddr_in *addr4;
+		struct sockaddr_in6 *addr6;
+	} _addr = {.addr = (struct sockaddr *)&addr_storage};
+	int proto = IPPROTO_TCP;
+
+#if defined(CONFIG_NET_SOCKETS_SOCKOPT_TLS)
+	proto = IPPROTO_TLS_1_2;
+#endif
+
+	HTTP_SERVICE_FOREACH(svc)
+	{
+		/* set the default address (in6addr_any / INADDR_ANY are all 0) */
+		addr_storage = (struct sockaddr_in6){0};
+
+		/* Set up the server address struct according to address family */
+		if (IS_ENABLED(CONFIG_NET_IPV6) &&
+		    inet_pton(AF_INET6, svc->host, &_addr.addr6->sin6_addr) == 1) {
+			/* if a literal IPv6 address is provided as the host, use IPv6 */
+			af = AF_INET6;
+			len = sizeof(*_addr.addr6);
+
+			_addr.addr6->sin6_family = AF_INET6;
+			_addr.addr6->sin6_port = *svc->port;
+		} else if (IS_ENABLED(CONFIG_NET_IPV4) &&
+			   inet_pton(AF_INET, svc->host, &_addr.addr4->sin_addr) == 1) {
+			/* if a literal IPv4 address is provided as the host, use IPv4 */
+			af = AF_INET;
+			len = sizeof(*_addr.addr4);
+
+			_addr.addr4->sin_family = AF_INET;
+			_addr.addr4->sin_port = *svc->port;
+		} else if (IS_ENABLED(CONFIG_NET_IPV6)) {
+			/* prefer IPv6 if both IPv6 and IPv4 are supported */
+			af = AF_INET6;
+			len = sizeof(*_addr.addr6);
+
+			_addr.addr6->sin6_family = AF_INET6;
+			_addr.addr6->sin6_port = *svc->port;
+		} else if (IS_ENABLED(CONFIG_NET_IPV4)) {
+			af = AF_INET;
+			len = sizeof(*_addr.addr4);
+
+			_addr.addr4->sin_family = AF_INET;
+			_addr.addr4->sin_port = *svc->port;
+		} else {
+			LOG_ERR("Neither IPv4 nor IPv6 is enabled");
+			break;
+		}
+		/* Create a socket */
+		ctx->server_fd = socket(af, SOCK_STREAM, proto);
```

Comment:
Isn't this an issue if there is more than 1 service? The `server_fd` member will have the value of the last socket, which is an issue for accepting new clients.